hack/build-image

```diff
@@ -0,0 +1,597 @@
+# of images we can build
+IMG_SERVER = "samba-server"
+IMG_AD_SERVER = "samba-ad-server"
+IMG_CLIENT = "samba-client"
+IMG_TOOLBOX = "samba-toolbox"
+IMAGE_KINDS = {
+    # short names
+    "server": IMG_SERVER,
+    "ad-server": IMG_AD_SERVER,
+    "client": IMG_CLIENT,
+    "toolbox": IMG_TOOLBOX,
+    # canonical names
+    "samba-server": IMG_SERVER,
+    "samba-ad-server": IMG_AD_SERVER,
+    "samba-client": IMG_CLIENT,
+    "samba-toolbox": IMG_TOOLBOX,
+}
+
+# ARCHITECTURES - map supported arch names/alias to canonical names
+AMD64 = "amd64"
+ARM64 = "arm64"
+ARCHITECTURES = {
+    # alternate names
+    "x86_64": AMD64,
+    "aarch64": ARM64,
+    # canonical names
+    "amd64": AMD64,
+    "arm64": ARM64,
+}
+
+# DISTROS - list of supported distro bases
+FEDORA = "fedora"
+CENTOS = "centos"
+OPENSUSE = "opensuse"
+DISTROS = [
+    FEDORA,
+    CENTOS,
+    OPENSUSE,
+]
+
+# PACKAGE_SOURCES - list of known package sources
+DEFAULT = "default"
+NIGHTLY = "nightly"
+PACKAGE_SOURCES = [DEFAULT, NIGHTLY]
+
+# SOURCE_DIRS - image source paths
+# (paths are relative to common image source dir)
+SOURCE_DIRS = {
+    IMG_SERVER: "server",
+    IMG_AD_SERVER: "ad-server",
+    IMG_CLIENT: "client",
+    IMG_TOOLBOX: "toolbox",
+}
+
+DEFAULT_PKG_SOURCES = [DEFAULT]
+DEFAULT_DISTRO_BASES = [FEDORA]
+LATEST = "latest"
+QUAL_NONE = "unqualified"
+QUAL_DISTRO = "distro-qualified"
+
+
+_DISCOVERED_CONTAINER_ENGINES = []
+
+
+def check_kind(kind):
+    """Return the canonical name for the image kind or raise a ValueError."""
+    try:
+        return IMAGE_KINDS[kind]
+    except KeyError:
+        raise ValueError(f"invalid kind: {kind}")
+
+
+def check_arch(arch):
+    """Return the canonical name for the arch or raise a ValueError."""
+    try:
+        return ARCHITECTURES[arch]
+    except KeyError:
+        raise ValueError(f"invalid arch: {arch}")
+
+
+def check_distro(distro):
+    """Return the canonical name for a distro base or raise a ValueError."""
+    if distro in DISTROS:
+        return distro
+    raise ValueError(f"invalid distro: {distro}")
+
+
+def check_pkg_source(source):
+    """Return the canonical name for a package source or raise a ValueError."""
+    if source in PACKAGE_SOURCES:
+        return source
+    raise ValueError(f"invalid package source: {source}")
+
+
+def check_repo_base_for(value):
+    """Return a tuple with a (distro_base, repo_base) pair or raise a
+    ValueError.
+    """
+    if "=" not in value:
+        raise ValueError("expected '=' in value")
+    db, rb = value.split("=", 1)
+    db = check_distro(db)
+    return (db, rb)
+
+
+def _cmd_to_str(cmd):
+    """Format a command for logging."""
+    return " ".join(shlex.quote(arg) for arg in cmd)
+
+
+def run(cli, cmd, capture_output=False, check=False):
+    """Execute a command. Wraps subprocess.run."""
+    if cli.dry_run and not capture_output:
+        logger.info("Would run: %s", _cmd_to_str(cmd))
+        return
```

Comment:
Mixed return type int this function. Better have `return ""` here for future type-hints annotations.

Reply:
Will update.